course9/lecture_en.md

```diff
@@ -0,0 +1,273 @@
+struct BoxedInt { value : Int } derive(Default, Eq, Compare, Debug)
+```
+
+## Example: Using Traits to Implement a Map
+
+Now, let's proceed to implement a generic map using traits.
+
+A map is a collection of key-value pairs, where each key is associated with a value. It is possible for different keys to correspond to the same value. For instance, in the map `{ 0 -> "a", 5 -> "Hello", 7 -> "a"}`, the key `5` corresponds to the value `"Hello"`, while both keys `0` and `7` correspond to the value `"a"`.
+
+```moonbit no-check
+type Map[Key, Value]
+```
+
+A map should support the following methods:
+- Create a map.
+  ```moonbit no-check
+  fn make[Key, Value]() -> Map[Key, Value]
+  ```
+- Add a key-value pair, or update the corresponding value of a key.
+  ```moonbit no-check
+  fn put[Key, Value](map: Map[Key, Value], key: Key, value: Value) -> Map[Key, Value]
+  ```
+- Get the corresponding value of a key.
+  ```moonbit no-check
+  fn get[Key, Value](map: Map[Key, Value], key: Key) -> Option[Value]
+  ```
+  Since such a key-value pair may not exist in the map, the return value is wrapped in `Option`.
+
+The map can be implemented using a list of pairs.
+
+```moonbit
+type Map[Key, Value] List[(Key, Value)]
+```
+
+The first two basic methods, `make` and `put`, can be easily implemented as follows:
+- Create a map by creating an empty list.
+  ```moonbit
+  fn make[Key, Value]() -> Map[Key, Value] { 
+    Map(Nil)
+  }
+  ```
+- Add/update a key-value pair by inserting the pair to the beginning of the list.
+  ```moonbit
+  fn put[Key, Value](map: Map[Key, Value], key: Key, value: Value) -> Map[Key, Value] { 
+    let Map(original_map) = map
+    Map( Cons( (key, value), original_map ) )
+  }
+  ```
+
+The third method, `get`, is also easy to describe in prose:
+- Search the list from the beginning until the first matching key is found.
+
+In such an implementation of the `get` function, we need to compare the key we are searching for with the keys stored in the map to determine if they are equal. As a result, the `Key` type must implement the `Eq` trait. That is, we need to modify the previous declaration of `get` so that `Key` is bound by `Eq`.
+
+The complete implementation of `get` is as follows:
+
+```moonbit
+fn get[Key: Eq, Value](map : Map[Key, Value], key : Key) -> Option[Value] {
+  fn aux(list : List[(Key, Value)]) -> Option[Value] {
+    match list {
+      Nil => None
+      Cons((k, v), tl) => if k == key {
+        // `Key` is bound by `Eq`, so we can call `==` directly.
+        Some(v)
+      } else {
+        aux(tl)
+      }
+    }
+  }
+
+  aux(map.0) // Use `.0` to get the value.
+}
+```
+
+In the above implementation, we define an auxiliary function called `aux` to traverse the list and use the `==` operator to determine whether the current key matches the one we are looking for. It can be shown that when there are multiple identical keys, we will always retrieve the latest value associated with that key.
```

Comment:
```suggestion
In the above implementation, we loop through the list and use the `==` operator to determine whether the current key matches the one we are looking for. It can be shown that when there are multiple identical keys, we will always retrieve the latest value associated with that key.
```